### PR TITLE
Cleanup `c/experimental/gradients/` part 3

### PR DESCRIPTION
**Pull request by vnghia**
@saxenasaurabh 
Could you take a look at this PR ? Thank you !

### CODE REVIEW

**Comment by vnvo2409** [comment]
@saxenasaurabh 

With a396295, each gradient in `math_grad` is tested and their tests ( as well as their helpers like `*GradModel` ) in other places ( `gradient_tests` ) are removed ( except `TestMatMulGrad`, please see my comment in the code ).

2 things I would like to discuss with you:
- With `gradient_checker`, I think we should compute `(f(x + eps) - f(x))/eps` instead of `(f(x + eps) - f(x - eps))/2*eps`. This way, we could avoid round-off errors when computing `2*eps` and `x-eps`.
- I think we will need one more PR for cleaning up `c/experimental/gradients` ( mostly `mnist_gradients_test` and friends ). We should start thinking what we gonna do with [unified_api_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/experimental/unified_api_test.py).

WDYT ?

File: tensorflow/c/eager/gradients_test.cc [hunk 1]
```diff
@@ -175,135 +93,6 @@ Status IdentityNGradModel(AbstractContext* ctx,
...
-                                           /*sources=*/inputs,
-                                           /*output_gradients=*/{}, outputs));
-  for (auto div_output : div_outputs) {
-    div_output->Unref();
-  }
-  delete tape;
-  return Status::OK();
-}
-
 Status getValue(AbstractTensorHandle* t, TF_Tensor** result_tensor) {
```
Comment (saxenasaurabh) [inline tensorflow/c/eager/gradients_test.cc:96]: Can we remove this and replace usage with tensorflow::gradients::GetValue?

File: tensorflow/c/experimental/gradients/grad_test_helper.cc [hunk 2]
```diff
@@ -105,6 +106,36 @@ void CheckTensorValue(AbstractTensorHandle* t, absl::Span<const float> manuals,
   delete[] danalytical;
 }
 
+Model BuildGradModel(Model ops, size_t num_inputs, string ops_name,
```
Comment (saxenasaurabh) [inline tensorflow/c/experimental/gradients/grad_test_helper.cc:109]: Do we need to provide `num_inputs` here or could we use `inputs.size()` ?

File: tensorflow/c/experimental/gradients/grad_test_helper.cc [hunk 3]
```diff
@@ -105,6 +106,36 @@ void CheckTensorValue(AbstractTensorHandle* t, absl::Span<const float> manuals,
...
+                     GradientFunctionFactory gradient_function_factory) {
```
Comment (saxenasaurabh) [inline tensorflow/c/experimental/gradients/grad_test_helper.cc:110]: The current signature restricts the usage of `BuildGradModel` to models containing a single op. A more general API might be `BuildGradModel(Model forward, const GradientRegistry& registry)`.

**Review by saxenasaurabh — COMMENTED** [review]
Thanks for cleaning this up. This is fantastic work!

**Comment by saxenasaurabh** [comment]
> @saxenasaurabh
> 
> With [a396295](https://github.com/tensorflow/tensorflow/commit/a3962951d81e08526edf6eef232d346148008af2), each gradient in `math_grad` is tested and their tests ( as well as their helpers like `*GradModel` ) in other places ( `gradient_tests` ) are removed ( except `TestMatMulGrad`, please see my comment in the code ).
> 
> 2 things I would like to discuss with you:
> 
> * With `gradient_checker`, I think we should compute `(f(x + eps) - f(x))/eps` instead of `(f(x + eps) - f(x - eps))/2*eps`. This way, we could avoid round-off errors when computing `2*eps` and `x-eps`.

Could you explain a bit more here? Right now this matches the logic in the [python gradient checker](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/ops/gradient_checker_v2.py;l=236;drc=1427bfc12ec5a3a2c6a4ffd57fc5b465d3eedfae)

> * I think we will need one more PR for cleaning up `c/experimental/gradients` ( mostly `mnist_gradients_test` and friends ). We should start thinking what we gonna do with [unified_api_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/experimental/unified_api_test.py).
> 
> WDYT ?

We could punt on cleaning that up. I am working on calling C++ gradient function from production python APIs. Once we have that we could simply delete the gradient tests here. The MNIST MLP benchmark is still useful. We should add a similar one for ResNet once we have all gradient functions for that.

**Comment by vnvo2409** [comment]
> Could you explain a bit more here?

You could see that `gradient_checker` does not work very well with `MatMul`. If we use `(f(x + eps) - f(x - eps))/2*eps`, we have to compute `a:=x+eps`, `b:=x-eps` and `c:=2*eps`. And because of the limitation of precision, `a` does not mathematically equal to `x+eps`, it is just an approximation. The same thing holds for `b` and `c`. So we have 3 sources of error. If we use `(f(x + eps) - f(x))/eps`, we only have 1 source.

I am not sure if I am missing something and why the `gradient_checker` in python is implemented that way 😕 

What do you think ?

File: tensorflow/c/eager/gradients_test.cc (hunk 1 above)
Comment (vnvo2409) [reply tensorflow/c/eager/gradients_test.cc]: Done

File: tensorflow/c/experimental/gradients/grad_test_helper.cc (hunk 3 above)
Comment (vnvo2409) [reply tensorflow/c/experimental/gradients/grad_test_helper.cc]: Done

File: tensorflow/c/experimental/gradients/grad_test_helper.cc (hunk 2 above)
Comment (vnvo2409) [reply tensorflow/c/experimental/gradients/grad_test_helper.cc]: Done

**Comment by vnvo2409** [comment]
@saxenasaurabh Could you take a look at this PR ? Thank you !

File: tensorflow/c/experimental/gradients/math_grad_test.cc [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+      ctx_.reset(ctx_raw);
+    }
+
+    // Computing numerical gradients with TensorFloat-32 is numerically
+    // unstable. Some forward pass tests also fail with TensorFloat-32 due to
+    // low tolerances
+    enable_tensor_float_32_execution(false);
+  }
+
+  AbstractContextPtr ctx_;
```
Comment (saxenasaurabh) [inline tensorflow/c/experimental/gradients/math_grad_test.cc:105]: Could we rename this to `immediate_execution_ctx_`?

File: tensorflow/c/experimental/gradients/math_grad_test.cc [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+  GradientRegistry registry_;
```
Comment (saxenasaurabh) [inline tensorflow/c/experimental/gradients/math_grad_test.cc:106]: Do we need to declare `registry_` and `status_` here or can these be local to the tests?

**Comment by saxenasaurabh** [comment]
> You could see that gradient_checker does not work very well with MatMul. If we use (f(x + eps) - f(x - eps))/2*eps, we have to compute a:=x+eps, b:=x-eps and c:=2*eps. And because of the limitation of precision, a does not mathematically equal to x+eps, it is just an approximation. The same thing holds for b and c. So we have 3 sources of error. If we use (f(x + eps) - f(x))/eps, we only have 1 source.

In general the [symmetric derivative](https://en.wikipedia.org/wiki/Symmetric_derivative) is a much better approximation of the derivative. To understand this you could try writing down the taylor series expansion which shows that the error for the symmetric derivative is `O(h^2)` compared to `O(h)` for the non-symmetric one.

File: tensorflow/c/experimental/gradients/math_grad_test.cc (hunk 5 above)
Comment (vnvo2409) [reply tensorflow/c/experimental/gradients/math_grad_test.cc:106]: they are indeed local to the tests. You could see that in `When these tests run, the following happens:` in https://github.com/google/googletest/blob/master/docs/primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests-same-data-multiple-tests

File: tensorflow/c/experimental/gradients/math_grad_test.cc (hunk 4 above)
Comment (vnvo2409) [reply tensorflow/c/experimental/gradients/math_grad_test.cc:105]: Done. In addition, this change and other changes will be applied to `nn_grad_test` in the next PR.

File: tensorflow/c/experimental/gradients/math_grad_test.cc (hunk 5 above)
Comment (saxenasaurabh) [reply tensorflow/c/experimental/gradients/math_grad_test.cc:106]: Yes, but I was wondering if we gain much from declaring these here?

Comment (vnvo2409) [reply tensorflow/c/experimental/gradients/math_grad_test.cc:106]: I think it saves us some lines ( with `registry_` ) or allows us to reuse the variable ( not really but with the previous code, we just created a `status`, used it once and threw it away. I think it is not really the best way ). WDYT ?

**Review by saxenasaurabh — APPROVED** [review]

**Comment by vnvo2409** [comment]
@saxenasaurabh 

Linux GPU is falling because `abs_error` is not big enough. Do you have any solution for it ?

**Comment by saxenasaurabh** [comment]
> @saxenasaurabh
> 
> Linux GPU is falling because `abs_error` is not big enough. Do you have any solution for it ?
> 
> Edit: I think `RunAndMaybeSum` might be the culprit. Do you think we should try to completely mimic the implementation of python (`_compute_theoretical_jacobian` and `_compute_numeric_jacobian` ) ?


Abs error of 0.4 seems too high. We should try to get to the bottom of this. If the python implementation for gradient checker is more accurate we should copy that. For now we could disable the Matmul test if that is only one failing and fix the gradient checker and enable this test in a follow up.

**Comment by vnvo2409** [comment]
> For now we could disable the Matmul test

Done

**Review by saxenasaurabh — APPROVED** [review]

**Comment by vnvo2409** [comment]
Linux GPU is falling because of unrelated tests so no need to worries.

**Pull request merged**